main/src/com/google/refine/model/Project.java

```diff
@@ -270,6 +280,80 @@ public void update() {
         columnModel.update();
         recordModel.update(this);
     }
+    
+    public void updateColumnChange(ColumnChange change) {
+        JSONObject schemaObj = null;
+        Schema schema = null;
+        
+        this.update();
+        // update the data package metadata
+        DataPackageMetadata dp = getDataPackageMetadata();
+        if (dp == null) {
+            dp = (DataPackageMetadata) MetadataFactory.buildMetadata(MetadataFormat.DATAPACKAGE_METADATA);
+            setMetadata(MetadataFormat.DATAPACKAGE_METADATA, dp);
+        }
+        
+        try {
+             if (dp.getPackage().getResources().size() > 0) {
+                 schemaObj = dp.getPackage().getResources().get(0).getSchema();
+                 schema = new Schema(schemaObj);
+             } else {
+                 Resource resource = SchemaExtension.createResource(getProjectMetadata().getName(),
+                         columnModel);
+                 dp.getPackage().addResource(resource);
+                 schema = new Schema();
+             }
+        } catch (Exception e) {
+            logger.error(ExceptionUtils.getStackTrace(e));
+        }
+        
+        String changeClazzName = change.getClass().getSimpleName();
+        switch (changeClazzName) {
```

Comment:
Sorry, but this is really not acceptable: it is not extensible at all. What if an extension declares a new operation that changes the columns? It will not be able to register the operation in your switch.

I am afraid it is going to be very hard to recover your implementation - the whole architecture needs to be started again from scratch. Again, the clean solution here is to store the schema in the `Column` objects themselves (or in a class that is linked directly from them).